File: packages/go/crudian/crud.go
package crudian

import (
	"context"
	"fmt"
)

// Crud is the table-oriented facade (JS createCrud equivalent).
type Crud struct {
	ex Executor
	d  Dialect
}

// NewCrud wires an executor and dialect.
func NewCrud(ex Executor, d Dialect) *Crud {
	if d == nil {
		d = SqliteDialect{}
	}
	return &Crud{ex: ex, d: d}
}

func (c *Crud) requireWhere(w *WhereBuilder, label string) error {
	if w == nil {
		return NewError(label + " requires where")
	}
	return nil
}

func (c *Crud) Create(ctx context.Context, table string, cols Row) (Row, error) {
	tbl, err := AssertString(table, "table")
	if err != nil {
		return nil, err
	}
	if cols == nil || len(cols) == 0 {
		return nil, NewError("cols must not be empty")
	}
	keys := sortedKeys(cols)
	qTbl := c.d.QuoteIdent(tbl)
	colSQL := make([]string, len(keys))
	ph := make([]string, len(keys))
	args := make([]any, len(keys))
	for i, k := range keys {
		colSQL[i] = c.d.QuoteIdent(k)
		ph[i] = c.d.Placeholder(i + 1)
		args[i] = cols[k]
	}
	sql := fmt.Sprintf(
		"INSERT INTO %s (%s) VALUES (%s) RETURNING *",
		qTbl, joinComma(colSQL), joinComma(ph),
	)
	row, err := c.ex.Get(ctx, sql, args...)
	if err != nil {
		return nil, err
	}
	if row == nil {
		return nil, NewError("expected row object")
	}
	return row, nil
}

func (c *Crud) Read(ctx context.Context, table string, query ReadQuery) (Row, error) {
	tbl, err := AssertString(table, "table")
	if err != nil {
		return nil, err
	}
	where, err := compileWhere(c.d, resolveWhere(query.Where))
	if err != nil {
		return nil, err
	}
	sql := "SELECT " + selectColumns(c.d, query.Columns) + " FROM " + c.d.QuoteIdent(tbl)
	if where.SQL != "" {
		sql += " WHERE " + where.SQL
	}
	sql += " LIMIT 1"
	return c.ex.Get(ctx, sql, where.Args...)
}

func (c *Crud) Update(ctx context.Context, table string, cols Row, query UpdateQuery) (Row, error) {
	tbl, err := AssertString(table, "table")
	if err != nil {
		return nil, err
	}
	if err := c.requireWhere(query.Where, "update"); err != nil {
		return nil, err
	}
	if cols == nil || len(cols) == 0 {
		return nil, NewError("cols must not be empty")
	}
	where, err := compileWhere(c.d, resolveWhere(query.Where))
	if err != nil {
		return nil, err
	}
	if where.SQL == "" {
		return nil, NewError("update requires where")
	}
	keys := sortedKeys(cols)
	sets := make([]string, len(keys))
	args := make([]any, 0, len(keys)+len(where.Args))
	for i, k := range keys {
		sets[i] = c.d.QuoteIdent(k) + " = " + c.d.Placeholder(i+1)
		args = append(args, cols[k])
	}
	args = append(args, where.Args...)
	qTbl := c.d.QuoteIdent(tbl)
	n, err := c.ex.Run(ctx, fmt.Sprintf("UPDATE %s SET %s WHERE %s", qTbl, joinComma(sets), where.SQL), args...)
	if err != nil {
		return nil, err
	}
	if n == 0 {
		return nil, nil
	}
	return c.ex.Get(ctx, "SELECT * FROM "+qTbl+" WHERE "+where.SQL+" LIMIT 1", where.Args...)
}

func (c *Crud) Delete(ctx context.Context, table string, query DeleteQuery) (int64, error) {
	tbl, err := AssertString(table, "table")
	if err != nil {
		return 0, err
	}
	if err := c.requireWhere(query.Where, "delete"); err != nil {
		return 0, err
	}
	where, err := compileWhere(c.d, resolveWhere(query.Where))
	if err != nil {
		return 0, err
	}
	if where.SQL == "" {
		return 0, NewError("delete requires where")
	}
	return c.ex.Run(ctx, "DELETE FROM "+c.d.QuoteIdent(tbl)+" WHERE "+where.SQL, where.Args...)
}

func (c *Crud) Count(ctx context.Context, table string, query CountQuery) (int64, error) {
	tbl, err := AssertString(table, "table")
	if err != nil {
		return 0, err
	}
	where, err := compileWhere(c.d, resolveWhere(query.Where))
	if err != nil {
		return 0, err
	}
	sql := "SELECT COUNT(*) AS " + c.d.QuoteIdent("row_count") + " FROM " + c.d.QuoteIdent(tbl)
	if where.SQL != "" {
		sql += " WHERE " + where.SQL
	}
	row, err := c.ex.Get(ctx, sql, where.Args...)
	if err != nil {
		return 0, err
	}
	if row == nil {
		return 0, nil
	}
	return toInt64(row["row_count"]), nil
}

func (c *Crud) Search(ctx context.Context, table string, query SearchQuery) (SearchResult, error) {
	tbl, err := AssertString(table, "table")
	if err != nil {
		return SearchResult{}, err
	}
	limit := query.Limit
	if limit == 0 {
		limit = 20
	}
	if limit <= 0 {
		return SearchResult{}, NewError("limit must be a positive number")
	}
	if query.Cursor != nil {
		switch query.Cursor.(type) {
		case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64, float32, float64, string:
		default:
			return SearchResult{}, NewError("cursor must be a number, string, or null")
		}
	}
	where, err := compileWhere(c.d, resolveWhere(query.Where))
	if err != nil {
		return SearchResult{}, err
	}
	args := append([]any{}, where.Args...)
	parts := make([]string, 0, 2)
	if where.SQL != "" {
		parts = append(parts, "("+where.SQL+")")
	}
	if query.Cursor != nil {
		parts = append(parts, c.d.QuoteIdent("id")+" > "+c.d.Placeholder(len(args)+1))
		args = append(args, query.Cursor)
	}
	whereSQL := ""
	if len(parts) > 0 {
		whereSQL = " WHERE " + parts[0]
		for i := 1; i < len(parts); i++ {
			whereSQL += " AND " + parts[i]
		}
	}
	sql := "SELECT " + selectColumns(c.d, query.Columns) + " FROM " + c.d.QuoteIdent(tbl) +
		whereSQL + " ORDER BY " + c.d.QuoteIdent("id") + " ASC LIMIT " + c.d.Placeholder(len(args)+1)
	args = append(args, limit+1)
	rows, err := c.ex.All(ctx, sql, args...)
	if err != nil {
		return SearchResult{}, err
	}
	hasMore := len(rows) > limit
	items := rows
	if hasMore {
		items = rows[:limit]
	}
	var next any
	if hasMore && len(items) > 0 {
		if id, ok := items[len(items)-1]["id"]; ok {
			next = id
		}
	}
	total, err := c.Count(ctx, table, CountQuery{Where: query.Where})
	if err != nil {
		return SearchResult{}, err
	}
	return SearchResult{Items: items, NextCursor: next, HasMore: hasMore, Total: total}, nil
}

func (c *Crud) List(ctx context.Context, table string, query SearchQuery) (SearchResult, error) {
	return c.Search(ctx, table, query)
}

func (c *Crud) Upsert(ctx context.Context, table string, cols Row) (Row, error) {
	tbl, err := AssertString(table, "table")
	if err != nil {
		return nil, err
	}
	if cols == nil || len(cols) == 0 {
		return nil, NewError("cols must not be empty")
	}
	id, ok := cols["id"]
	if !ok {
		return nil, NewError("upsert requires cols.id")
	}
	existing, err := c.Read(ctx, tbl, ReadQuery{Where: Where().Eq("id", id)})
	if err != nil {
		return nil, err
	}
	if existing != nil {
		patch := Row{}
		for k, v := range cols {
			if k == "id" {
				continue
			}
			patch[k] = v
		}
		if len(patch) == 0 {
			return existing, nil
		}
		updated, err := c.Update(ctx, tbl, patch, UpdateQuery{Where: Where().Eq("id", id)})
		if err != nil {
			return nil, err
		}
		if updated == nil {
			return nil, NewError("upsert update failed")
		}
		return updated, nil
	}
	return c.Create(ctx, tbl, cols)
}

func (c *Crud) Duplicate(ctx context.Context, table string, query DuplicateQuery) (Row, error) {
	tbl, err := AssertString(table, "table")
	if err != nil {
		return nil, err
	}
	if err := c.requireWhere(query.Where, "duplicate"); err != nil {
		return nil, err
	}
	source, err := c.Read(ctx, tbl, ReadQuery{Where: query.Where})
	if err != nil || source == nil {
		return nil, err
	}
	cols := Row{}
	for k, v := range source {
		if k == "id" {
			continue
		}
		cols[k] = v
	}
	for k, v := range query.Overrides {
		cols[k] = v
	}
	delete(cols, "id")
	return c.Create(ctx, tbl, cols)
}

func (c *Crud) BulkCreate(ctx context.Context, table string, rows []Row) (int64, error) {
	if len(rows) == 0 {
		return 0, nil
	}
	var n int64
	for _, row := range rows {
		if _, err := c.Create(ctx, table, row); err != nil {
			return n, err
		}
		n++
	}
	return n, nil
}

func (c *Crud) BulkUpdate(ctx context.Context, table string, cols Row, query UpdateQuery) (int64, error) {
	tbl, err := AssertString(table, "table")
	if err != nil {
		return 0, err
	}
	if err := c.requireWhere(query.Where, "bulkUpdate"); err != nil {
		return 0, err
	}
	if cols == nil || len(cols) == 0 {
		return 0, NewError("cols must not be empty")
	}
	where, err := compileWhere(c.d, resolveWhere(query.Where))
	if err != nil {
		return 0, err
	}
	if where.SQL == "" {
		return 0, NewError("bulkUpdate requires where")
	}
	keys := sortedKeys(cols)
	sets := make([]string, len(keys))
	args := make([]any, 0, len(keys)+len(where.Args))
	for i, k := range keys {
		sets[i] = c.d.QuoteIdent(k) + " = " + c.d.Placeholder(i+1)
		args = append(args, cols[k])
	}
	args = append(args, where.Args...)
	return c.ex.Run(ctx, fmt.Sprintf("UPDATE %s SET %s WHERE %s", c.d.QuoteIdent(tbl), joinComma(sets), where.SQL), args...)
}

func (c *Crud) BulkDelete(ctx context.Context, table string, query DeleteQuery) (int64, error) {
	return c.Delete(ctx, table, query)
}

func (c *Crud) BulkUpsert(ctx context.Context, table string, rows []Row) (int64, error) {
	if len(rows) == 0 {
		return 0, nil
	}
	var n int64
	for _, row := range rows {
		if row == nil {
			return n, NewError("each row must be an object")
		}
		if _, ok := row["id"]; !ok {
			return n, NewError("bulkUpsert requires each row to have id")
		}
		if _, err := c.Upsert(ctx, table, row); err != nil {
			return n, err
		}
		n++
	}
	return n, nil
}

func (c *Crud) Transaction(ctx context.Context, fn func(tx *Crud) error) error {
	if fn == nil {
		return NewError("transaction callback must be a function")
	}
	return c.ex.Transaction(ctx, func(txEx Executor) error {
		return fn(NewCrud(txEx, c.d))
	})
}

func sortedKeys(cols Row) []string {
	keys := make([]string, 0, len(cols))
	for k := range cols {
		keys = append(keys, k)
	}
	// stable-ish for SQL: sort
	for i := 0; i < len(keys); i++ {
		for j := i + 1; j < len(keys); j++ {
			if keys[j] < keys[i] {
				keys[i], keys[j] = keys[j], keys[i]
			}
		}
	}
	return keys
}

func toInt64(v any) int64 {
	switch n := v.(type) {
	case int64:
		return n
	case int:
		return int64(n)
	case int32:
		return int64(n)
	case float64:
		return int64(n)
	case float32:
		return int64(n)
	default:
		return 0
	}
}

File: packages/php/src/Crud.php
<?php

declare(strict_types=1);

namespace B4moss\Crudian;

final class Crud
{
    /** @var array<string, true> */
    private array $pkOK = [];

    public function __construct(
        private readonly Executor $ex,
        private readonly Dialect $d,
        private readonly string $pk,
    ) {
    }

    public static function newWith(Executor $ex, ?Dialect $d = null, ?Options $options = null): self
    {
        $dialect = $d ?? resolveOptionsDialect($options);

        return new self($ex, $dialect, resolvePk($options));
    }

    private function ensurePkColumn(string $table): void
    {
        if (isset($this->pkOK[$table])) {
            return;
        }
        [$sql, $args] = $this->d->describeColumns($table);
        $rows = $this->ex->all($sql, $args);
        $found = false;
        foreach ($rows as $row) {
            $name = $row['name'] ?? $row['NAME'] ?? null;
            if ($name === $this->pk) {
                $found = true;
                break;
            }
        }
        if (!$found) {
            throw CrudianError::of('pk column "' . $this->pk . '" does not exist on table "' . $table . '"');
        }
        $this->pkOK[$table] = true;
    }

    private function requireWhere(?WhereBuilder $w, string $label): void
    {
        if ($w === null) {
            throw CrudianError::of($label . ' requires where');
        }
    }

    /**
     * @param array<string, mixed> $cols
     * @return array<string, mixed>
     */
    public function create(string $table, array $cols): array
    {
        $tbl = assertNonEmptyString($table, 'table');
        $this->ensurePkColumn($tbl);
        if ($cols === []) {
            throw CrudianError::of('cols must not be empty');
        }
        $keys = sortedKeys($cols);
        $qTbl = $this->d->quoteIdent($tbl);
        $colSQL = [];
        $ph = [];
        $args = [];
        $i = 1;
        foreach ($keys as $k) {
            $colSQL[] = $this->d->quoteIdent($k);
            $ph[] = $this->d->placeholder($i);
            $args[] = $cols[$k];
            $i++;
        }
        if ($this->d->supportsInsertReturning()) {
            $sql = sprintf(
                'INSERT INTO %s (%s) VALUES (%s) RETURNING *',
                $qTbl,
                joinComma($colSQL),
                joinComma($ph),
            );
            $row = $this->ex->get($sql, $args);
            if ($row === null) {
                throw CrudianError::of('expected row object');
            }

            return $row;
        }
        $this->ex->run(
            sprintf('INSERT INTO %s (%s) VALUES (%s)', $qTbl, joinComma($colSQL), joinComma($ph)),
            $args,
        );
        if (array_key_exists($this->pk, $cols)) {
            $pkValue = $cols[$this->pk];
        } else {
            $idRow = $this->ex->get($this->d->lastInsertIdSql());
            if ($idRow === null) {
                throw CrudianError::of('expected row object');
            }
            $pkValue = $idRow['id'] ?? null;
        }
        $row = $this->ex->get(
            'SELECT * FROM ' . $qTbl . ' WHERE ' . $this->d->quoteIdent($this->pk) . ' = ' . $this->d->placeholder(1),
            [$pkValue],
        );
        if ($row === null) {
            throw CrudianError::of('expected row object');
        }

        return $row;
    }

    /**
     * @param array{where?: ?WhereBuilder, columns?: list<string>} $query
     * @return array<string, mixed>|null
     */
    public function read(string $table, array $query = []): ?array
    {
        $tbl = assertNonEmptyString($table, 'table');
        $this->ensurePkColumn($tbl);
        $where = compileWhere($this->d, resolveWhere($query['where'] ?? null), 1);
        $sql = 'SELECT ' . selectColumns($this->d, $query['columns'] ?? []) . ' FROM ' . $this->d->quoteIdent($tbl);
        if ($where->sql !== '') {
            $sql .= ' WHERE ' . $where->sql;
        }
        $sql .= ' LIMIT 1';

        return $this->ex->get($sql, $where->args);
    }

    /**
     * @param array<string, mixed> $cols
     * @param array{where?: ?WhereBuilder} $query
     * @return array<string, mixed>|null
     */
    public function update(string $table, array $cols, array $query = []): ?array
    {
        $tbl = assertNonEmptyString($table, 'table');
        $this->ensurePkColumn($tbl);
        $this->requireWhere($query['where'] ?? null, 'update');
        if ($cols === []) {
            throw CrudianError::of('cols must not be empty');
        }
        $keys = sortedKeys($cols);
        $sets = [];
        $args = [];
        $i = 1;
        foreach ($keys as $k) {
            $sets[] = $this->d->quoteIdent($k) . ' = ' . $this->d->placeholder($i);
            $args[] = $cols[$k];
            $i++;
        }
        $where = compileWhere($this->d, resolveWhere($query['where'] ?? null), count($keys) + 1);
        if ($where->sql === '') {
            throw CrudianError::of('update requires where');
        }
        $args = [...$args, ...$where->args];
        $qTbl = $this->d->quoteIdent($tbl);
        $n = $this->ex->run(
            sprintf('UPDATE %s SET %s WHERE %s', $qTbl, joinComma($sets), $where->sql),
            $args,
        );
        if ($n === 0) {
            return null;
        }
        $fetchWhere = compileWhere($this->d, resolveWhere($query['where'] ?? null), 1);

        return $this->ex->get(
            'SELECT * FROM ' . $qTbl . ' WHERE ' . $fetchWhere->sql . ' LIMIT 1',
            $fetchWhere->args,
        );
    }

    /**
     * @param array{where?: ?WhereBuilder} $query
     */
    public function delete(string $table, array $query = []): int
    {
        $tbl = assertNonEmptyString($table, 'table');
        $this->ensurePkColumn($tbl);
        $this->requireWhere($query['where'] ?? null, 'delete');
        $where = compileWhere($this->d, resolveWhere($query['where'] ?? null), 1);
        if ($where->sql === '') {
            throw CrudianError::of('delete requires where');
        }

        return $this->ex->run(
            'DELETE FROM ' . $this->d->quoteIdent($tbl) . ' WHERE ' . $where->sql,
            $where->args,
        );
    }

    /**
     * @param array{where?: ?WhereBuilder} $query
     */
    public function count(string $table, array $query = []): int
    {
        $tbl = assertNonEmptyString($table, 'table');
        $this->ensurePkColumn($tbl);
        $where = compileWhere($this->d, resolveWhere($query['where'] ?? null), 1);
        $sql = 'SELECT COUNT(*) AS ' . $this->d->quoteIdent('row_count') . ' FROM ' . $this->d->quoteIdent($tbl);
        if ($where->sql !== '') {
            $sql .= ' WHERE ' . $where->sql;
        }
        $row = $this->ex->get($sql, $where->args);
        if ($row === null) {
            return 0;
        }

        return toInt($row['row_count'] ?? $row['ROW_COUNT'] ?? 0);
    }

    /**
     * @param array{where?: ?WhereBuilder} $query
     */
    public function exists(string $table, array $query = []): bool
    {
        $tbl = assertNonEmptyString($table, 'table');
        $this->ensurePkColumn($tbl);
        $where = compileWhere($this->d, resolveWhere($query['where'] ?? null), 1);
        $sql = 'SELECT 1 AS ' . $this->d->quoteIdent('ok') . ' FROM ' . $this->d->quoteIdent($tbl);
        if ($where->sql !== '') {
            $sql .= ' WHERE ' . $where->sql;
        }
        $sql .= ' LIMIT 1';

        return $this->ex->get($sql, $where->args) !== null;
    }

    /**
     * @param array{
     *   where?: ?WhereBuilder,
     *   columns?: list<string>,
     *   limit?: int,
     *   cursor?: mixed,
     *   paging?: string,
     *   offset?: int|null
     * } $query
     */
    public function search(string $table, array $query = []): SearchResult
    {
        $tbl = assertNonEmptyString($table, 'table');
        $this->ensurePkColumn($tbl);
        $limit = $query['limit'] ?? 0;
        if ($limit === 0) {
            $limit = 20;
        }
        if ($limit <= 0) {
            throw CrudianError::of('limit must be a positive number');
        }

        $paging = $query['paging'] ?? '';
        if ($paging === '') {
            $paging = 'offset';
        }
        if ($paging !== 'offset' && $paging !== 'cursor') {
            throw CrudianError::of('paging must be "offset" or "cursor"');
        }
        $cursor = $query['cursor'] ?? null;
        $offsetOpt = array_key_exists('offset', $query) ? $query['offset'] : null;
        if ($paging === 'offset' && $cursor !== null) {
            throw CrudianError::of('offset paging does not accept cursor');
        }
        if ($paging === 'cursor' && $offsetOpt !== null) {
            throw CrudianError::of('cursor paging does not accept offset');
        }
        if ($cursor !== null && !is_int($cursor) && !is_float($cursor) && !is_string($cursor)) {
            throw CrudianError::of('cursor must be a number, string, or null');
        }

        $where = compileWhere($this->d, resolveWhere($query['where'] ?? null), 1);
        $total = $this->count($tbl, ['where' => $query['where'] ?? null]);

        if ($paging === 'offset') {
            $offset = is_int($offsetOpt) ? $offsetOpt : 0;
            if ($offset < 0) {
                throw CrudianError::of('offset must be a non-negative number');
            }
            $args = $where->args;
            $whereSQL = $where->sql !== '' ? ' WHERE ' . $where->sql : '';
            $idx = $where->nextIndex;
            $sql = 'SELECT ' . selectColumns($this->d, $query['columns'] ?? []) . ' FROM ' . $this->d->quoteIdent($tbl)
                . $whereSQL . ' ORDER BY ' . $this->d->quoteIdent($this->pk) . ' ASC LIMIT ' . $this->d->placeholder($idx)
                . ' OFFSET ' . $this->d->placeholder($idx + 1);
            $args[] = $limit;
            $args[] = $offset;
            $rows = $this->ex->all($sql, $args);

            return new SearchResult(
                items: $rows,
                hasMore: ($offset + count($rows)) < $total,
                total: $total,
                offset: $offset,
                limit: $limit,
            );
        }

        $args = $where->args;
        $parts = [];
        if ($where->sql !== '') {
            $parts[] = '(' . $where->sql . ')';
        }
        $idx = $where->nextIndex;
        if ($cursor !== null) {
            $parts[] = $this->d->quoteIdent($this->pk) . ' > ' . $this->d->placeholder($idx);
            $args[] = $cursor;
            $idx++;
        }
        $whereSQL = '';
        if ($parts !== []) {
            $whereSQL = ' WHERE ' . implode(' AND ', $parts);
        }
        $sql = 'SELECT ' . selectColumns($this->d, $query['columns'] ?? []) . ' FROM ' . $this->d->quoteIdent($tbl)
            . $whereSQL . ' ORDER BY ' . $this->d->quoteIdent($this->pk) . ' ASC LIMIT ' . $this->d->placeholder($idx);
        $args[] = $limit + 1;
        $rows = $this->ex->all($sql, $args);
        $hasMore = count($rows) > $limit;
        $items = $hasMore ? array_slice($rows, 0, $limit) : $rows;
        $next = null;
        if ($hasMore) {
            $last = $items[array_key_last($items)];
            $next = $last[$this->pk] ?? null;
        }

        return new SearchResult(
            items: $items,
            hasMore: $hasMore,
            total: $total,
            nextCursor: $next,
        );
    }

    /**
     * @param array{
     *   where?: ?WhereBuilder,
     *   columns?: list<string>,
     *   limit?: int,
     *   cursor?: mixed,
     *   paging?: string,
     *   offset?: int|null
     * } $query
     */
    public function list(string $table, array $query = []): SearchResult
    {
        return $this->search($table, $query);
    }

    /**
     * @param array<string, mixed> $cols
     * @return array<string, mixed>
     */
    public function upsert(string $table, array $cols): array
    {
        $tbl = assertNonEmptyString($table, 'table');
        $this->ensurePkColumn($tbl);
        if ($cols === []) {
            throw CrudianError::of('cols must not be empty');
        }
        if (!array_key_exists($this->pk, $cols)) {
            throw CrudianError::of('upsert requires cols.' . $this->pk);
        }
        $id = $cols[$this->pk];
        $existing = $this->read($tbl, ['where' => WhereBuilder::create()->eq($this->pk, $id)]);
        if ($existing !== null) {
            $patch = [];
            foreach ($cols as $k => $v) {
                if ($k === $this->pk) {
                    continue;
                }
                $patch[$k] = $v;
            }
            if ($patch === []) {
                return $existing;
            }
            $updated = $this->update($tbl, $patch, ['where' => WhereBuilder::create()->eq($this->pk, $id)]);
            if ($updated === null) {
                throw CrudianError::of('upsert update failed');
            }

            return $updated;
        }

        return $this->create($tbl, $cols);
    }

    /**
     * @param array{where?: ?WhereBuilder, overrides?: array<string, mixed>} $query
     * @return array<string, mixed>|null
     */
    public function duplicate(string $table, array $query = []): ?array
    {
        $tbl = assertNonEmptyString($table, 'table');
        $this->ensurePkColumn($tbl);
        $this->requireWhere($query['where'] ?? null, 'duplicate');
        $source = $this->read($tbl, ['where' => $query['where'] ?? null]);
        if ($source === null) {
            return null;
        }
        $cols = [];
        foreach ($source as $k => $v) {
            if ($k === $this->pk) {
                continue;
            }
            $cols[$k] = $v;
        }
        foreach ($query['overrides'] ?? [] as $k => $v) {
            $cols[$k] = $v;
        }
        unset($cols[$this->pk]);

        return $this->create($tbl, $cols);
    }

    /**
     * @param list<array<string, mixed>> $rows
     */
    public function bulkCreate(string $table, array $rows): int
    {
        if ($rows === []) {
            return 0;
        }
        $n = 0;
        foreach ($rows as $row) {
            $this->create($table, $row);
            $n++;
        }

        return $n;
    }

    /**
     * @param array<string, mixed> $cols
     * @param array{where?: ?WhereBuilder} $query
     */
    public function bulkUpdate(string $table, array $cols, array $query = []): int
    {
        $tbl = assertNonEmptyString($table, 'table');
        $this->ensurePkColumn($tbl);
        $this->requireWhere($query['where'] ?? null, 'bulkUpdate');
        if ($cols === []) {
            throw CrudianError::of('cols must not be empty');
        }
        $keys = sortedKeys($cols);
        $sets = [];
        $args = [];
        $i = 1;
        foreach ($keys as $k) {
            $sets[] = $this->d->quoteIdent($k) . ' = ' . $this->d->placeholder($i);
            $args[] = $cols[$k];
            $i++;
        }
        $where = compileWhere($this->d, resolveWhere($query['where'] ?? null), count($keys) + 1);
        if ($where->sql === '') {
            throw CrudianError::of('bulkUpdate requires where');
        }
        $args = [...$args, ...$where->args];

        return $this->ex->run(
            sprintf('UPDATE %s SET %s WHERE %s', $this->d->quoteIdent($tbl), joinComma($sets), $where->sql),
            $args,
        );
    }

    /**
     * @param array{where?: ?WhereBuilder} $query
     */
    public function bulkDelete(string $table, array $query = []): int
    {
        return $this->delete($table, $query);
    }

    /**
     * @param list<array<string, mixed>> $rows
     */
    public function bulkUpsert(string $table, array $rows): int
    {
        if ($rows === []) {
            return 0;
        }
        $n = 0;
        foreach ($rows as $row) {
            if (!array_key_exists($this->pk, $row)) {
                throw CrudianError::of('bulkUpsert requires each row to have ' . $this->pk);
            }
            $this->upsert($table, $row);
            $n++;
        }

        return $n;
    }

    /**
     * @param callable(self): void $fn
     */
    public function transaction(callable $fn): void
    {
        $this->ex->transaction(function (Executor $txEx) use ($fn): void {
            $tx = new self($txEx, $this->d, $this->pk);
            $tx->pkOK = $this->pkOK;
            $fn($tx);
        });
    }
}
